src/cape_ros_node.cpp

```diff
@@ -0,0 +1,337 @@
+          ros::topic::waitForMessage<sensor_msgs::CameraInfo>("depth/camera_info", ros::Duration(1.0));
+      if (cam_intrinsics_info)
+      {
+        if (cam_intrinsics_info->K[0] > 0 && cam_intrinsics_info->K[4] > 0 && cam_intrinsics_info->K[2] > 0 &&
+            cam_intrinsics_info->K[5] > 0)
+        {
+          fx_ = cam_intrinsics_info->K[0];
+          fy_ = cam_intrinsics_info->K[4];
+          cx_ = cam_intrinsics_info->K[2];
+          cy_ = cam_intrinsics_info->K[5];
+          intrinsics_ready_ = true;
+        }
+      }
+      else
+      {
+        ROS_ERROR("Failed to obtain camera intrinsic messages");
+        return;
+      }
+      ROS_WARN("fx, fy, px, py: %f, %f, %f, %f", fx_, fy_, cx_, cy_);
+    }
+    catch (std::exception e)
+    {
+      ROS_ERROR("Failed to obtain camera intrinsic messages: %s", e.what());
+      return;
+    }
+  }
+
+  cv_bridge::CvImageConstPtr image_ptr = cv_bridge::toCvShare(image, image->encoding);
+  auto image_depth16 = reinterpret_cast<const uint16_t*>(image_ptr->image.data);
+  float scaled_depth(0.);
+
+  if (!cape_extractor_ || height_ != image->height || width_ != image->width)
+  {
+    height_ = image->height;
+    width_ = image->width;
+    reset();
+  }
+
+  auto iter_mat = depth_mat_.begin<float>();
+
+  // Go through the depth iamge and scale each pixel
+  for (int y = 0; y < height_; ++y)
+  {
+    for (int x = 0; x < width_; ++x)
+    {
+      scaled_depth = static_cast<float>(*image_depth16) * depth_scale_;
+
+      // Check if the point is valid
+      if (scaled_depth <= 0.f)
+      {
+        *iter_mat = 0.;
+      }
+      else
+        *iter_mat = scaled_depth;
+
+      ++image_depth16;
+      ++iter_mat;
+    }
+  }
+
+  X_ = X_pre_.mul(depth_mat_);
+  Y_ = Y_pre_.mul(depth_mat_);
+  cloud_array_.setZero();
+  seg_output_.setTo(0);
+  plane_params_.clear();
+  cylinder_params_.clear();
+
+  simpleProjectPointCloud(X_, Y_, depth_mat_, cloud_array_, depth_cutoff_);
+
+  organizePointCloudByCell(cloud_array_, cloud_array_organized_, cell_map_);
+
+  cape_extractor_->process(cloud_array_organized_, seg_output_, plane_params_, cylinder_params_);
+
+  planes_pub_.publish(generateMessage(image->header));
+
+  if (intensity_image_ptr_ && (image_overlay_pub_.getNumSubscribers() > 0))
+  {
+    overlaySegmentsOnImage(intensity_image_ptr_->image, seg_output_);
+    image_overlay_pub_.publish(cv_bridge::CvImage(image->header, "rgb8", overlay_image_).toImageMsg());
+  }
+}
+
+void CapeRosNode::reset()
+{
+  cape_extractor_ = std::make_shared<Cape>(height_, width_, patch_size_, patch_size_, cylinder_detection_,
+                                           cos_angle_max_, max_merge_dist_);
+  depth_mat_.create(height_, width_, CV_32F);
+  X_.create(height_, width_, CV_32F);
+  Y_.create(height_, width_, CV_32F);
+  X_pre_.create(height_, width_, CV_32F);
+  Y_pre_.create(height_, width_, CV_32F);
+
+  cloud_array_.resize(width_ * height_, 3);
```

Comment:
Can this be height_ * width_ instead? I know it doesn't make a difference here, just for consistency can we always use these in the order of height then width. Same goes for the line below, too.